src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs

```diff
@@ -8,16 +8,117 @@ namespace Microsoft.Azure.ServiceBus.Core
+            string entityPath,
+            ReceiveMode receiveMode = ReceiveMode.PeekLock,
+            RetryPolicy retryPolicy = null)
+            : this(entityPath, new ServiceBusNamespaceConnection(connectionString), receiveMode, retryPolicy)
+        {
+            if (string.IsNullOrWhiteSpace(connectionString))
+            {
+                throw Fx.Exception.ArgumentNullOrWhiteSpace(connectionString);
+            }
+            if (string.IsNullOrWhiteSpace(entityPath))
+            {
+                throw Fx.Exception.ArgumentNullOrWhiteSpace(entityPath);
+            }
+
+            this.ownsConnection = true;
+        }
+
+        public MessageReceiver(
+            string entityPath,
+            ServiceBusConnection serviceBusConnection,
+            ReceiveMode receiveMode = ReceiveMode.PeekLock,
+            RetryPolicy retryPolicy = null)
+            : base(nameof(MessageSender) + StringUtility.GetRandomString(), retryPolicy ?? RetryPolicy.Default)
+        {
+            this.ReceiveMode = receiveMode;
+            this.operationTimeout = serviceBusConnection.OperationTimeout;
+            this.messageReceivePumpSyncLock = new object();
+            this.entityPath = entityPath;
+            this.ServiceBusConnection = serviceBusConnection;
+            this.PrefetchCount = prefetchCount;
+            var tokenProvider = TokenProvider.CreateSharedAccessSignatureTokenProvider(serviceBusConnection.SasKeyName, serviceBusConnection.SasKey);
+            this.CbsTokenProvider = new TokenProviderAdapter(tokenProvider, serviceBusConnection.OperationTimeout);
+            this.ReceiveLinkManager = new FaultTolerantAmqpObject<ReceivingAmqpLink>(this.CreateLinkAsync, this.CloseSession);
+            this.RequestResponseLinkManager = new FaultTolerantAmqpObject<RequestResponseAmqpLink>(this.CreateRequestResponseLinkAsync, this.CloseRequestResponseSession);
+            this.requestResponseLockedMessages = new ConcurrentExpiringSet<Guid>();
+        }
+
+        internal MessageReceiver(
+            string entityPath,
+            MessagingEntityType? entityType,
+            ReceiveMode receiveMode,
+            int prefetchCount,
+            ServiceBusConnection serviceBusConnection,
+            ICbsTokenProvider cbsTokenProvider,
+            RetryPolicy retryPolicy)
+            : this(entityPath, entityType, receiveMode, prefetchCount, serviceBusConnection, cbsTokenProvider, null, retryPolicy)
+        {
+        }
+
+        internal MessageReceiver(
+            string entityPath,
+            MessagingEntityType? entityType,
+            ReceiveMode receiveMode,
+            int prefetchCount,
+            ServiceBusConnection serviceBusConnection,
+            ICbsTokenProvider cbsTokenProvider,
+            string sessionId,
+            RetryPolicy retryPolicy,
+            bool isSessionReceiver = false)
+            : base(nameof(MessageReceiver) + StringUtility.GetRandomString(), retryPolicy ?? RetryPolicy.Default)
```

Comment:
>: base(nameof(MessageReceiver) + StringUtility.GetRandomString(), retryPolicy ?? RetryPolicy.Default) [](start = 12, length = 101)

as commented earlier, let all the constructors finally reach this constructor and then this could do `: base()`